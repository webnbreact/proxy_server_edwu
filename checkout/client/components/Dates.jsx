import React from 'react';
import moment from 'moment';
import { BoldText, BoxWrapper, TimeAgoStyle } from './styled/Styled.jsx';

import 'react-dates/initialize';
import 'react-dates/lib/css/_datepicker.css';
import { DateRangePicker } from 'react-dates';
import './styled/react_dates_overrides.css';

class Dates extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      startDate: null,
      endDate: null,
      datesCompleted: false,
      focusedInput: null,
      unavailableDays: [],
    };
    this.isDayBlocked = this.isDayBlocked.bind(this);
  }

  componentDidUpdate() {
    const { bookedDates } = this.props;
    const unavailableDays = [];
    if (this.state.unavailableDays.length === 0) {
      for (let i = 0; i < bookedDates.length; i += 1) {
        let startDay = moment(bookedDates[i].check_in);
        let endDay = moment(bookedDates[i].check_out);
        for (let currDay = startDay; currDay.isBefore(endDay); currDay.add(1, 'day')) {
          unavailableDays.push(currDay.format());
        }
      }
      this.setState({ unavailableDays });
    }
    const { startDate, endDate, datesCompleted } = this.state;
    if (startDate !== null && endDate !== null && !datesCompleted) {
      this.setState({ datesCompleted: true }, () => {
        this.props.handleDateChange(startDate, endDate);
      });
    }
    if (startDate === null && endDate === null && datesCompleted) {
      this.setState( { datesCompleted: false }, () => {
        this.props.handleDateChange(null, null);
      });
    }
  }

  isDayBlocked(day) {
    return this.state.unavailableDays.some((unavailableDay) => {
      return moment(unavailableDay).isSame(day, 'day');
    });
  }

  render() {
    const timeAgo = moment(this.props.lastUpdated).fromNow();
    return (
      <div>
        <div style={{ marginBottom: '5px' }}>
          <BoldText>Dates</BoldText>
        </div>
        <BoxWrapper>
          <DateRangePicker
            startDateId="checkinDate"
            endDateId="checkoutDate"
            startDatePlaceholderText="Check in"
            endDatePlaceholderText="Check out"
            startDate={this.state.startDate}
            endDate={this.state.endDate}
            numberOfMonths={1}
            onDatesChange={({ startDate, endDate }) => this.setState({ startDate, endDate })}
            focusedInput={this.state.focusedInput}
            onFocusChange={focusedInput => this.setState({ focusedInput })}
            readOnly={true}
            showClearDates={true}
            noBorder={true}
            customCloseIcon={<i className="far fa-calendar-times clear-cal" />}
            renderCalendarInfo={() => (<TimeAgoStyle>Updated {timeAgo}</TimeAgoStyle>)}
            isDayBlocked={this.isDayBlocked}
          />
        </BoxWrapper>
      </div>
    );
  }
}

export default Dates;
