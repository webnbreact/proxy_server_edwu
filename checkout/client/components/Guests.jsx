import React from 'react';
import GuestDropdown from './GuestDropdown.jsx';
import { GuestsStyle, BoldText, BoxText, BoxIcon, BoxWrapper, GuestMenuBtn } from './styled/Styled.jsx';

const Guests = ({guests, dropdown, displayDropdown, handleGuestCountChange}) => {
  const guestCount = guests.adults + guests.children;
  const infantCount = guests.infants;
  const infantDisplay = (<span>, {infantCount} infant{infantCount > 1 ? 's' : null}</span>);
  const dropdownToDisplay = (dropdown === 'guest') ? 'none' : 'guest';
  return (
    <GuestsStyle>
      <div style={{ marginBottom: '5px' }}>
        <BoldText>Guests</BoldText>
      </div>
      <BoxWrapper>
        <GuestMenuBtn type="button" onClick={() => displayDropdown(dropdownToDisplay)}>
          <div>
            <BoxText dropdown={dropdown}>
              <div>
                <span>
                  {guestCount} guest{guestCount > 1 ? 's' : null}
                </span>
                {infantCount > 0 ? infantDisplay : null}
              </div>
            </BoxText>
            <BoxIcon>
              {(dropdown === 'guest') ? <i className="fas fa-angle-up" style={{ fontSize: '1.4em' }} />
                : <i className="fas fa-angle-down" style={{ fontSize: '1.4em' }} />}
            </BoxIcon>
          </div>
        </GuestMenuBtn>
        {(dropdown === 'guest') ? <GuestDropdown guests={guests} displayDropdown={displayDropdown} handleGuestCountChange={handleGuestCountChange} /> : null}
      </BoxWrapper>
    </GuestsStyle>
  );
}

export default Guests;
