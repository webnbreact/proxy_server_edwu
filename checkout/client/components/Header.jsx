import React from 'react';
import { Button, BoldText, Price, StarStyle } from './styled/Styled.jsx';

const Header = ({ state }) => {
  const numStars = Math.floor(state.avgRating / 2);
  const numHalfStars = state.avgRating % 2;
  const star = <i className="fas fa-star" />;
  const halfStar = <i className="fas fa-star-half" />;

  return (
    <div className="override-line-height" style={{ marginBottom: '15px' }}>
      <div>
        <Price>${state.nightlyPrice}</Price>
        <BoldText> per night</BoldText>
      </div>
      <div className="override-line-height">
        <Button>
          <span>
            {[...Array(numStars)].map((item, index) => {
              return (
                <StarStyle key={index}>
                  {star}
                </StarStyle>
              );
            })}
            {[...Array(numHalfStars)].map((item, index) => {
              return (
                <StarStyle key={index + 5}>
                  {halfStar}
                </StarStyle>
              );
            })}
            <BoldText style={{left: '3px', fontSize: '12px'}}>{state.reviewCount}</BoldText>
          </span>
        </Button>
      </div>
    </div>
  );
};

export default Header;
