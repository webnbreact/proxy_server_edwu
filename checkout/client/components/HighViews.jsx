import React from 'react';
import {
  Divider, FlexBox, BoldText, Text, SlideDown,
} from './styled/Styled.jsx';

const HighViews = ({ pastWeekViews }) => (
  <SlideDown>
      <Divider style={{ margin: '20px 0' }} />
      <FlexBox style={{ justifyContent: 'space-between', flexWrap: 'wrap' }}>
        <div style={{ marginLeft: '1px' }}>
          <BoldText style={{ fontSize: '14.5px' }}>
            This home is on people's minds.
          </BoldText>
          <Text>
            It's been viewed { pastWeekViews } times in the past
            <br />
            week.
          </Text>
        </div>
        <div style={{ fontSize: '2.25em', padding: '8px', textShadow: '#f0bb3d 1px 0 10px' }}>
          <i className="far fa-lightbulb" />
        </div>
      </FlexBox>
  </SlideDown>
);

export default HighViews;
