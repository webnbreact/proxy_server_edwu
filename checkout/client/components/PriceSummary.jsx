import React from 'react';
import {
  PriceSummDivider, PriceSummaryStyle, PriceSummaryLine, ButtonQuest,
} from './styled/Styled.jsx';

const PriceSummary = ({ nights, nightlyPrice, cleaningFee, serviceFee }) => (
  <PriceSummaryStyle>
    <PriceSummaryLine>
      <div>
        <span>${nightlyPrice} x {nights} night{nights > 1 ? 's' : null}</span>
      </div>
      <div>
        <span>${nightlyPrice * nights}</span>
      </div>
    </PriceSummaryLine>
    <PriceSummDivider />
    <PriceSummaryLine>
      <div>
        <span>Cleaning fee</span>
        <ButtonQuest
          type="button"
          data-toggle="tooltip"
          title="One-time fee charged by host to cover the cost of cleaning their space."
        >
          <i className="far fa-question-circle" />
        </ButtonQuest>
      </div>
      <div>
        <span>${cleaningFee}</span>
      </div>
    </PriceSummaryLine>
    <PriceSummDivider />
    <PriceSummaryLine>
      <div>
        <span className="service-fee">Service fee</span>
        <ButtonQuest
          type="button"
          data-toggle="tooltip"
          title="This helps us run our platform and offer services like 24/7 support on your trip."
        >
          <i className="far fa-question-circle" />
        </ButtonQuest>
      </div>
      <div>
        <span>${serviceFee * nights}</span>
      </div>
    </PriceSummaryLine>
    <PriceSummDivider />
    <PriceSummaryLine style={{ fontWeight: '600' }}>
      <div>
        <span className="total">Total</span>
      </div>
      <div>
        <span>${(nightlyPrice + serviceFee) * nights + cleaningFee}</span>
      </div>
    </PriceSummaryLine>
  </PriceSummaryStyle>
);

export default PriceSummary;
