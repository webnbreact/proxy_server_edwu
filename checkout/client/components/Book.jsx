import React from 'react';
import $ from 'jquery';
import Dates from './Dates.jsx';
import Guests from './Guests.jsx';
import PriceSummary from './PriceSummary.jsx';
import { ButtonSubmit, CenterText, BookStyle } from './styled/Styled.jsx';

class Book extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      // checkedIn: false,    // default value should be false
      // checkedOut: false,   // changed to true for testing
      checkinDate: null,
      checkoutDate: null,
      dropdown: 'none',     // can be 'guest', 'checkinCal', or 'checkoutCal'
      guests: {
        adults: 1,
        children: 0,
        infants: 0,
      },
    };
    this.displayDropdown = this.displayDropdown.bind(this);
    this.handleGuestCountChange = this.handleGuestCountChange.bind(this);
    this.handleDateChange = this.handleDateChange.bind(this);
  }

  componentDidUpdate() {
    $('[data-toggle=tooltip]')
      .tooltip({
        trigger: 'focus',
      });
  }

  displayDropdown(item) {
    this.setState({
      dropdown: item,
    });
  }

  handleDateChange(startDate, endDate) {
    this.setState({
      checkinDate: startDate,
      checkoutDate: endDate,
    });
  }

  handleGuestCountChange(ageGroup, count) {
    const {adults, children, infants} = this.state.guests;
    const totalGuests = adults + children;
    if (this.state.guests[ageGroup] === 0 && count === -1) {
      return;
    }
    if (totalGuests >= 4 && count === 1 && ageGroup !== 'infants') {
      return;
    }
    if (infants >= 5 && ageGroup === 'infants' && count === 1) {
      return;
    }
    const guests = {...this.state.guests};
    guests[ageGroup] = this.state.guests[ageGroup] + count;
    this.setState({guests});
  }

  render() {
    const { nightlyPrice, cleaningFee, serviceFee, lastUpdated, bookedDates } = this.props.state;
    const { checkinDate, checkoutDate } = this.state;
    let nights = 0;
    if (checkinDate !== null && checkoutDate !== null) {
      nights = checkoutDate.diff(checkinDate, 'days');
    }
    return (
      <BookStyle>
        <form action="">
          <div className="override-line-height">
            <Dates lastUpdated={lastUpdated} bookedDates={bookedDates} handleDateChange={this.handleDateChange} />
            <Guests guests={this.state.guests} dropdown={this.state.dropdown} displayDropdown={this.displayDropdown} handleGuestCountChange={this.handleGuestCountChange} />
          </div>
          <div>
            {checkinDate !== null && checkoutDate !== null
              ? <PriceSummary nights={nights} nightlyPrice={nightlyPrice} cleaningFee={cleaningFee} serviceFee={serviceFee} /> : null}
          </div>
          <div>
            <ButtonSubmit type="submit">
              <span>Request to Book</span>
            </ButtonSubmit>
          </div>
          <div>
            <CenterText>You won't be charged yet</CenterText>
          </div>
        </form>
      </BookStyle>
    );
  }
}

export default Book;
