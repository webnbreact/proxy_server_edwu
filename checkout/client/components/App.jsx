import React from 'react';
import Header from './Header.jsx';
import Book from './Book.jsx';
import HighViews from './HighViews.jsx';
import {
  GlobalStyle, Module, MainWrapper, FlagWidget, ButtonUnderline, Divider,
} from './styled/Styled.jsx';

class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      roomId: '',
      nightlyPrice: '',
      pastWeekViews: '',
      cleaningFee: '',
      serviceFee: '',
      lastUpdated: '',
      reviewCount: '',
      avgRating: '',
      bookedDates: [],
    };
  }

  componentDidMount() {
    const roomId = window.location.pathname.slice(7, -1);
    const randomRoomId = Math.ceil(Math.random() * 100);
    fetch(`http://localhost:1337/rooms/${roomId || randomRoomId}/booking`)
      .then((data) => {
        return data.json();
      }).then((roomInfo) => {
        const { room_id, nightly_price, past_week_views, cleaning_fee, service_fee, last_updated, review_count, avg_rating } = roomInfo[0];
        const bookedDates = [];
        for (let i = 0; i < roomInfo.length; i += 1) {
          const { check_in, check_out } = roomInfo[i];
          bookedDates.push({ check_in, check_out });
        }
        this.setState({
          roomId: room_id,
          nightlyPrice: nightly_price,
          pastWeekViews: past_week_views,
          cleaningFee: cleaning_fee,
          serviceFee: service_fee,
          lastUpdated: last_updated,
          reviewCount: review_count,
          avgRating: avg_rating,
          bookedDates: bookedDates,
        });
      }).catch((err) => { throw err; });
  }

  render() {
    let views = this.state.pastWeekViews;
    views = (views >= 500) ? '500+' : views;
    return (
      <MainWrapper>
        <GlobalStyle />
        <Module>
          <div>
            <div>
              <Header state={this.state} />
              <Divider />
              <Book state={this.state} />
            </div>
          </div>
          <div>
            {this.state.pastWeekViews > 150 ? <HighViews pastWeekViews={views} /> : null}
          </div>
        </Module>
        <FlagWidget>
          <ButtonUnderline>
            <i className="far fa-flag" />
            <span> Report this listing</span>
          </ButtonUnderline>
        </FlagWidget>
      </MainWrapper>
    );
  }
}

export default App;
