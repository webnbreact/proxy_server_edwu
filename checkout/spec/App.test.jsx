import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import App from '../client/components/App.jsx';
import Header from '../client/components/Header.jsx';
import Book from '../client/components/Book.jsx';
import Dates from '../client/components/Dates.jsx';
import Guests from '../client/components/Guests.jsx';
import PriceSummary from '../client/components/PriceSummary.jsx';
import HighViews from '../client/components/HighViews.jsx';
import {
  Price, Button, ButtonSubmit, ButtonQuest, BoldText, Text, GuestMenuBtn,
} from '../client/components/styled/Styled.jsx';

const sinon = require('sinon');
const { sampleState } = require('./sampleData.js');

describe('App component', () => {
  const wrapper = shallow(<App />, { disableLifecycleMethods: true });
  
  it('should render without crashing', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should store state for the room\'s price and fees', () => {
    const state = wrapper.state();
    expect(state.nightlyPrice).toBeDefined();
    expect(state.cleaningFee).toBeDefined();
    expect(state.serviceFee).toBeDefined();
  });

  it('should render the Header and Book components', () => {
    expect(wrapper.find(Header)).toHaveLength(1);
    expect(wrapper.find(Book)).toHaveLength(1);
  });
});

describe('Header component', () => {
  const wrapper = mount(<Header state={sampleState} />);

  it('should contain a valid number for the nightly price', () => {
    const price = wrapper.find(Price).text();
    const priceDigits = price.slice(1);
    expect(Number(priceDigits)).not.toBe(NaN);
  });

  it('should contain a button that indicates the number of reviews', () => {
    const reviewsButton = wrapper.find(Button);
    const reviewCount = wrapper.find(BoldText).get(1).props.children;
    expect(reviewsButton).toHaveLength(1);
    expect(typeof reviewCount).toBe('number');
  });
});

describe('Book component', () => {
  const wrapper = mount(<Book state={sampleState} />);

  it('should include a "Request to Book" button', () => {
    const bookingButton = wrapper.find(ButtonSubmit);
    expect(bookingButton).toHaveLength(1);
  });

  it('should not display the Price Summary on initial render', () => {
    const priceSumm = wrapper.find(PriceSummary);
    expect(priceSumm).toHaveLength(0);
  });

  it('should render the Dates and Guests components', () => {
    expect(wrapper.find(Dates)).toHaveLength(1);
    expect(wrapper.find(Guests)).toHaveLength(1);
  });
});

describe('PriceSummary component', () => {
  const wrapper = mount(<PriceSummary />);

  it('should include a line for the Service Fee and Total', () => {
    expect(wrapper.find('span.service-fee').text()).toBe('Service fee');
    expect(wrapper.find('span.total').text()).toBe('Total');
  });

  it('should contain at least one help button for explaining the \
  service fee (cleaning fee is not always charged)', () => {
    expect(wrapper.find(ButtonQuest).length).toBeGreaterThanOrEqual(1);
  });

  it('should render in the correct format', () => {
    const shallowWrapper = shallow(<PriceSummary />);
    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });
});

describe('Guests component', () => {
  const displayDropdownFake = sinon.fake();
  const guests = {};  // guests prop is empty because it's not used in the tests
  const wrapper = mount(<Guests guests={guests} displayDropdown={displayDropdownFake} />);

  it('should invoke the displayDropdown method when GuestMenuBtn is clicked', () => {
    wrapper.find(GuestMenuBtn).simulate('click');
    expect(displayDropdownFake.callCount).toBe(1);
  });

  it('should display the text "Guests" at the top of the component', () => {
    const title = wrapper.find(BoldText).text();
    expect(title).toBe('Guests');
  });
});

describe('HighViews component', () => {
  it('should contain the text "This home is on people\'s minds"', () => {
    const wrapper = shallow(<HighViews />);
    const text = wrapper.find(BoldText).text();
    expect(text).toBe('This home is on people\'s minds.');
  });

  it('should display "500+ times" when the past week view count is over 500', () => {
    let viewCount1 = 499;
    viewCount1 = (viewCount1 >= 500) ? '500+' : viewCount1; // logic taken from App component
    const wrapper1 = mount(<HighViews pastWeekViews={viewCount1} />);
    const viewedText1 = wrapper1.find(Text).text();
    expect(viewedText1).not.toBe('It\'s been viewed 500+ times in the pastweek.');
    let viewCount2 = 501;
    viewCount2 = (viewCount2 >= 500) ? '500+' : viewCount2; // logic taken from App component
    const wrapper2 = mount(<HighViews pastWeekViews={viewCount2} />);
    const viewedText2 = wrapper2.find(Text).text();
    expect(viewedText2).toBe('It\'s been viewed 500+ times in the pastweek.');
  });
});
