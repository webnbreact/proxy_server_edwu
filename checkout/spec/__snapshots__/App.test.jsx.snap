// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should render without crashing 1`] = `
<styled.div>
  <GlobalStyleComponent />
  <styled.div>
    <div>
      <div>
        <Header
          state={
            Object {
              "avgRating": "",
              "bookedDates": Array [],
              "cleaningFee": "",
              "lastUpdated": "",
              "nightlyPrice": "",
              "pastWeekViews": "",
              "reviewCount": "",
              "roomId": "",
              "serviceFee": "",
            }
          }
        />
        <styled.hr />
        <Book
          state={
            Object {
              "avgRating": "",
              "bookedDates": Array [],
              "cleaningFee": "",
              "lastUpdated": "",
              "nightlyPrice": "",
              "pastWeekViews": "",
              "reviewCount": "",
              "roomId": "",
              "serviceFee": "",
            }
          }
        />
      </div>
    </div>
    <div />
  </styled.div>
  <styled.div>
    <Styled(styled.button)>
      <i
        className="far fa-flag"
      />
      <span>
         Report this listing
      </span>
    </Styled(styled.button)>
  </styled.div>
</styled.div>
`;

exports[`PriceSummary component should render in the correct format 1`] = `
<styled.div>
  <styled.div>
    <div>
      <span>
        $
         x 
         night
      </span>
    </div>
    <div>
      <span>
        $
        NaN
      </span>
    </div>
  </styled.div>
  <Styled(styled.hr) />
  <styled.div>
    <div>
      <span>
        Cleaning fee
      </span>
      <Styled(styled.button)
        data-toggle="tooltip"
        title="One-time fee charged by host to cover the cost of cleaning their space."
        type="button"
      >
        <i
          className="far fa-question-circle"
        />
      </Styled(styled.button)>
    </div>
    <div>
      <span>
        $
      </span>
    </div>
  </styled.div>
  <Styled(styled.hr) />
  <styled.div>
    <div>
      <span
        className="service-fee"
      >
        Service fee
      </span>
      <Styled(styled.button)
        data-toggle="tooltip"
        title="This helps us run our platform and offer services like 24/7 support on your trip."
        type="button"
      >
        <i
          className="far fa-question-circle"
        />
      </Styled(styled.button)>
    </div>
    <div>
      <span>
        $
        NaN
      </span>
    </div>
  </styled.div>
  <Styled(styled.hr) />
  <styled.div
    style={
      Object {
        "fontWeight": "600",
      }
    }
  >
    <div>
      <span
        className="total"
      >
        Total
      </span>
    </div>
    <div>
      <span>
        $
        NaN
      </span>
    </div>
  </styled.div>
</styled.div>
`;
